keps/sig-windows/2802-identify-windows-pods-apiserver-admission/README.md

```diff
@@ -0,0 +1,352 @@
+worst case, don't apply security constraints at all.
+
+
+### Goals
+
+Admission plugins and/or Admission Webhooks identifying if the incoming pod
+is Windows or not.
+
+
+### Non-Goals
+
+
+- Validating if the Windows pods have appropriate security constraints.
+- Interaction with container runtimes.
+
+## Proposal
+
+
+As mentioned [earlier](#Motivation) identifying the Windows pods during the scheduling phase can be done
+in the following ways:
+- Based on [nodeSelector and tolerations](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec with 
+  Windows node specific labels.
+- Based on [runtimeclasses](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/#ensuring-os-specific-workloads-land-on-the-appropriate-container-host) in the pod spec
+
+The problem with using nodeSelector and tolerations is that any unprivileged user/entity can apply the 
+nodeSelector and/or tolerations in the pod spec making it unsecure whereas RuntimeClasses are [recommended](https://kubernetes.io/docs/concepts/containers/runtime-class/#setup) to be
+created by cluster administrator making them authoritative enough to be used.
+
+In order to make `kubernetes.io/os` authoritative enough, kubelet should also reject admitting pods if the pod contains a nodeSelector with label `kubenetes.io/os` with a value not matching `runtime.GOOS`. We can use this check as `kubernetes.io/os` value is supposed to be filled from `runtime.GOOS`. Having this check in kubelet helps in the following conditions:
+- If scheduler has been by passed by setting the nodeName directly on the pod spec.
+- A malicious user or a user unknowingly with RBAC to just bind pod to node, can associate pod to node there by bypassing the scheduler.
+
+
+### User Stories (Optional)
+
+
+#### Story 1
+
+As a Kubernetes cluster administrator, I want appropriate security contexts to be 
+applied to my Windows Pods along with Linux pods
+
+#### Story 2
+
+As a Kubernetes cluster administrator, I want to use my own admission webhook for
+Windows pods.
+
+### Notes/Constraints/Caveats (Optional)
+
+While the `scheduling` field in the RuntimeClass is created to handle scheduling constraints, we have a reserved label
+[`kubernetes.io/os`](https://kubernetes.io/docs/reference/labels-annotations-taints/#kubernetes-io-os) to identify
+Windows nodes from their Linux counterparts. If the RuntimClass's scheduling field has a nodeSelector  `kubernetes.io/os: 'windows'`,
+we can authoritatively say that the pod is indeed a Windows one during the api-server admission time. Querying the RuntimeClass
+during admission time may be expensive operation as the admission plugin is not stateless anymore. 
+
+### Risks and Mitigations
+
+
+
+Existing users may be impacted as we're enforcing to use RuntimeClasses instead of plain nodeSelector and tolerations.
```

Comment:
breaking existing successful Pod API users in this way doesn't match my expectations for GA features

Reply:
specifically, this proposed change does not seem necessary or backwards compatible:

> Deny admission to pods that have only `kubernetes.io/os: windows` in nodeSelector but don't have an associated RuntimeClass with `kubernetes.io/os: windows` in scheduling field.

why is it necessary to reject such a pod? won't it get scheduled and run correctly?

Reply:
One scenario, that I noticed frequently is users specify a wrong nodeSelector, for example provide windows nodeSelector on linux host and vice-versa. If it is created via a workload controller which creates n replicas, since kubelet would allow such pods to be run the container runtime is going to throw the error and chances of node going unready are high. I just wanted to make sure that we're failing early.

Having said that, if the kubelet can rely on the nodeSelector of the pod and compare it with runtime.GOOS to admit/deny the pod, I think we are in a better shape than we were earlier and I believe it is ok to let users specify the nodeSelector and be less prescriptive.

Reply:
we would still have that pod in a constant bind->admit->reject loop, right?

similar to @liggitt though, i do not know how we can require usage of a runtime class.  for example, if a cluster had windows only nodes, control plane externalized, wouldn't this break backward compatibility for that cluster user?  who is responsible for creation of the runtime class?